### PR TITLE
selfcheck: Enable the `redundant-expr` error code

### PR DESCRIPTION
**Pull request by AlexWaygood**

### CODE REVIEW

File: mypy/build.py [hunk 1]
```diff
@@ -1292,13 +1292,7 @@ def find_cache_meta(id: str, path: str, manager: BuildManager) -> CacheMeta | No
     )
 
     # Don't check for path match, that is dealt with in validate_meta().
-    if (
-        m.id != id
-        or m.mtime is None
```
Comment (JelleZijlstra) [inline mypy/build.py:1297]: Not sure about this. Sure, these attributes may never be None according to their types, but it's good to be defensive against a corrupted cache.

File: mypy/fastparse.py [hunk 2]
```diff
@@ -905,7 +905,7 @@ def do_func_def(
                         self.errors, line=lineno, override_column=n.col_offset
                     ).translate_expr_list(func_type_ast.argtypes)
                     arg_types = [
-                        a if a is not None else AnyType(TypeOfAny.unannotated)
+                        a if a is not None else AnyType(TypeOfAny.unannotated)  # type: ignore[redundant-expr]
```
Comment (JelleZijlstra) [inline mypy/fastparse.py:908]: Why the ignore?

Comment (AlexWaygood) [reply mypy/fastparse.py]: I ran into annoying errors due to `list` invariance. But I think I've found a workaround (see 4def89f)

File: mypy/build.py (hunk 1 above)
Comment (AlexWaygood) [reply mypy/build.py:1297]: I see your point. I think this ultimately points to the annotations for `CacheMeta` being incorrect (or at least less than ideal). But changing those is somewhat of a large task, so I've added a TODO for now.

File: mypy/fastparse.py (hunk 2 above)
Comment (AlexWaygood) [reply mypy/fastparse.py]: Actually I changed my mind about 4def86f66aea836b64f908c460678366299a1906; I think 63815f00f66f20c0cf947e03a70e47324c5cb988 is a better solution to the problem

File: mypy/fastparse.py [hunk 3]
```diff
@@ -1042,7 +1040,7 @@ def transform_args(
         for a, kd in zip(args.kwonlyargs, args.kw_defaults):
             new_args.append(
                 self.make_argument(
-                    a, kd, ARG_NAMED if kd is None else ARG_NAMED_OPT, no_type_check
+                    a, kd, ARG_NAMED if kd is None else ARG_NAMED_OPT, no_type_check  # type: ignore[redundant-expr]
```
Comment (sobolevn) [inline mypy/fastparse.py:1043]: This one does not look right: `kd` can be `None`, can't it?

**Review by sobolevn — COMMENTED** [review]
Everything - except one thing - looks good to me!

File: mypy/fastparse.py (hunk 3 above)
Comment (JelleZijlstra) [reply mypy/fastparse.py:1043]: This isn't ast though. It's wrong in the `typed-ast` stubs: https://github.com/python/typeshed/blob/c0f27e39f9d424f9075b100f5c323f8422797ac9/stubs/typed-ast/typed_ast/ast3.pyi#L361

Comment (AlexWaygood) [reply mypy/fastparse.py:1043]: Oh, of course! Mypy uses stdlib `ast` on 3.8+, but uses `typed-ast` for Python 3.7, and the selfcheck runs with `--python-version=3.7`.

Comment (AlexWaygood) [reply mypy/fastparse.py:1043]: Typeshed PR here: https://github.com/python/typeshed/pull/8645.

**Review by sobolevn — APPROVED** [review]
Thanks!

**Pull request merged**